### PR TITLE
Automate testing of examples

### LINKED ISSUE #699: examples/tsp.jl is not tested
Bonjour,
Some months ago, I've succesfully tested `tsp.jl` with julia-0.3.1 and cplex or glpk (although the example is provide for gurobi).
I'm now testing the new tsp.jl exemple with the last JuMP version and julia-0.4.0 but get some error.
I saw [#684](https://github.com/JuliaOpt/JuMP.jl/issues/684), but it seems to be fixed!

Are code in the `examples` directory supposed to be maintained with each JuMP versions?

-- Maurice

<pre>
----
Inside subtour callback
Current tour starting at city 1:
[1,2,3,1]Adding subtour elimination cut
----
ERROR: LoadError: MethodError: `isless` has no method matching isless(::JuMP.GenericAffExpr{Float64,JuMP.Variable}, ::Int64)
Closest candidates are:
  isless(!Matched::AbstractFloat, ::Real)
  isless(!Matched::ForwardDiff.ForwardDiffNumber{N,T<:Real,C}, ::Real)
  isless(!Matched::Real, ::Real)
  ...
 in subtour at /..../run_tsp_cplex_distrib_20160310.jl:212
 in lazycallback at /.../pkgs/v0.4/JuMP/src/callbacks.jl:65
...
</pre>

### LINKED ISSUE #1175: Update examples for JuMP/MOI and format them as doctests
Using [doctests](https://juliadocs.github.io/Documenter.jl/stable/man/doctests/) would make sure the examples keep working over time. This addresses #699.

There are a couple open questions here.

- What solvers should we use for testing?
- We just went through a lot of work to make JuMP testable without solvers.... but using doctests would force us to depend on solvers in CI. A few options are:
  - Collapse JuMP/MOI/MOIU/MOIT into a single package and include one default solver.
  - Make a separate package called JuMPExamples.

I'm not entirely serious about collapsing JuMP/MOI/MOIU/MOIT into a single package, but if you look at any other modeling package like cvxpy, pyomo, etc., JuMP is an outlier in its modularity, and it results in weird situations like not having a solver available to run examples. I expect we'll go with JuMPExamples.

### PR DESCRIPTION
**Pull request by odow**
This PR adds a build stage to Linux/Julia 1.0 on Travis that runs the examples in `/examples`. This requires the installation of some solvers. So far, I have added GLPK (for LP/MIP), Ipopt (for NLP), and SCS (for SDP).

List of examples and whether they have been updated:
- [x] basic
- [x] cannery
- [x] cluster
- [x] corr_sdp
- [x] diet
- [x] knapsack
- [x] maxcut_sdp
- [x] mindistortion
- [x] minellipse
- [x] mle
- [x] multi
- [x] optcontrol
- [x] prod
- [x] qcp
- [x] robust_uncertainty
- [x] rosenbrock
- [x] steelT3
- [x] sudoku
- [x] transp
- [x] urbanplan

Closes #699 
Closes #1175 

- [x] PR currently includes `Manifest.toml` because latest GLPK tag isn't in METADATA. Once it is, we can remove `Manifest.tom`.

### CODE REVIEW

File: examples/basic.jl [hunk 1]
```diff
@@ -7,29 +7,47 @@
...
-#  st 1x + 5y <= 3
-#      0 <= x <= 2
-#      0 <= y <= 30
-#############################################################################
 
-using JuMP, Clp
+using JuMP, GLPK, Test
+
+#=
+    example_basic([; verbose = true])
```
Comment (mlubin) [inline examples/basic.jl]: Why not format as a docstring?

File: examples/mindistortion.jl [hunk 2]
```diff
@@ -3,58 +3,58 @@
...
-        @constraint(m, D[i,j]^2 <= Q[i,i] + Q[j,j] - 2Q[i,j])
-        @constraint(m, Q[i,i] + Q[j,j] - 2Q[i,j] <= D[i,j]^2*cSq )
-    end
-end
 
-@objective(m, Min, cSq)
+using JuMP, SCS, Test
+
+"""
+    example_mindistortion()
```
Comment (mlubin) [inline examples/mindistortion.jl]: Should be `example_min_distortion`, and the filename `min_distortion.jl`.

File: examples/mindistortion.jl [hunk 3]
```diff
@@ -3,58 +3,58 @@
...
+
+This example arises from computational geometry, in particular the problem of
+embedding a general finite metric space into a euclidean space.
+
+It is known that the 4-point metric space defined by the star graph:
+  x
+   \\
```
Comment (mlubin) [inline examples/mindistortion.jl]: Extra "\\"

File: examples/minellipse.jl [hunk 4]
```diff
@@ -3,52 +3,44 @@
...
+  subject to  X >= A_i,    i = 1,...,m
+              X PSD
+where W is a PD matrix of weights to choose between different solutions.
+"""
+function example_minellipse()
+    # We will use three ellipses: two "simple" ones, and a random one.
+    As = [
+        [2.0  0.0; 0.0  1.0],
+        [1.0  0.0; 0.0  3.0]  #,
+        # [2.86715 1.60645; 1.60645 1.12639]
```
Comment (mlubin) [inline examples/minellipse.jl]: Why is this line commented out?

File: examples/minellipse.jl [hunk 5]
```diff
@@ -3,52 +3,44 @@
...
 
-using JuMP
-using SCS
-using LinearAlgebra
-#using PyPlot  # Comment out if not installed
+using JuMP, SCS, LinearAlgebra, Test
 
-# solver = SCSSolver(eps=1e-6)
+"""
+    example_minellipse()
```
Comment (mlubin) [inline examples/minellipse.jl]: `example_min_ellipse` and `min_ellipse.jl`

File: examples/run_examples.jl [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    "qcp.jl",
+    "robust_uncertainty.jl",
+    "rosenbrock.jl",
+    "sudoku.jl",
+    "urbanplan.jl"
+]
+
+@testset "run_examples.jl" begin
+    for example in EXAMPLES
+        @testset "$(example)" begin
```
Comment (mlubin) [inline examples/run_examples.jl]: A more compact form is `@testset "$example" for example in EXAMPLES`.

File: examples/urbanplan.jl [hunk 7]
```diff
@@ -7,78 +7,68 @@
...
     # x is indexed by row and column
-    @variable(m, 0 <= x[1:5, 1:5] <= 1, Int)
+    @variable(model, 0 <= x[1:5, 1:5] <= 1, Int)
 
-    # y is indexed by R or C, and the points
-    # JuMP allows indexing on arbitrary sets
-    rowcol = ["R","C"]
-    points = [+5,+4,+3,-3,-4,-5]
-    @variable(m, 0 <= y[rowcol, points, 1:5] <= 1, Int)
+    # y is indexed by R or C, and the points JuMP allows indexing on arbitrary
```
Comment (mlubin) [inline examples/urbanplan.jl]: This sentence doesn't make sense. I think they were separate sentences before.

File: examples/mindistortion.jl (hunk 3 above)
Comment (odow) [reply examples/mindistortion.jl]: Because it is in the docstring it needed to escape the `\`.

**Comment by chriscoey** [comment]
Perhaps for a separate PR: can I suggest a renaming of some examples to further improve consistency (eg underscores) and understandability. I think we can be verbose with names here.

Eg steelT3 - no idea what the T3 means, off the bat

**Comment by hochB** [comment]
I want to say that for newcomers examples are more helpful than anything else.
So a big thank you for working on this.

**Comment by odow** [comment]
> Perhaps for a separate PR: can I suggest a renaming of some examples to further improve consistency (eg underscores) and understandability. I think we can be verbose with names here.

Totally agree. I'm almost verging on these examples should be included in the documentation and structured as tutorial/notebooky type things. But let's get them tidied up and passing tests first.

**Comment by odow** [comment]
Okay, all the examples have been updated to the latest JuMP master and are tested in a separate build stage on travis.

While I have tidied up many of the examples for style and more JuMPish conventions (see `diet.jl` whic uses `DenseAxisArray` to the data :)), I didn't fix all of them (e.g. `prod.jl`). 

Once this PR is in, I might move a few of the examples into the docs with doctests. Maybe one MIP, one NLP, and one conic.

File: examples/run_examples.jl [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+#  obtain one at http://mozilla.org/MPL/2.0/.
+
+#=
+    This script runs all the JuMP examples.
+    julia --project=examples run_examples.jl
+=#
+
+using Test
+
+const EXAMPLES = [
```
Comment (mlubin) [inline examples/run_examples.jl]: To get rid of an extra list that needs to be maintained, we could also run all `.jl` files in the examples directory.

File: examples/sudoku.jl [hunk 9]
```diff
@@ -1,4 +1,4 @@
-#  Copyright 2017, Iain Dunning, Joey Huchette, Miles Lubin, and contributors
+#  Copyright 2017 Iain Dunning, Joey Huchette Miles Lubin, and contributors
```
[the remark this answers is not in the document]
Comment (mlubin) [reply examples/sudoku.jl]: Seconded

File: examples/basic.jl [hunk 10]
```diff
@@ -4,7 +4,7 @@
 #  file, You can obtain one at http://mozilla.org/MPL/2.0/.
 #############################################################################
 # JuMP
-# An algebraic modeling langauge for Julia
+# An algebraic modeling language for Julia
```
[the remark this answers is not in the document]
Comment (odow) [reply examples/basic.jl]: See #1720

**Pull request merged**